### DISCUSSION TITLE
PDF / Finding checked and unchecked checkboxes

### DISCUSSION THREAD

**Discussion by Rapid1898-code** [jsvine/pdfplumber#782] [Ask for help with specific PDFs]
Hello, i try to check the pdf for checked or unchecked checkboxes - 

I use the following code to do this:

```
import pdfplumber
import os
import sys

if __name__ == '__main__':
  path = os.path.abspath(os.path.dirname(sys.argv[0])) 
  fn = os.path.join(path, "sample2.pdf")
  pdf = pdfplumber.open(fn)
  page = pdf.pages[0]
  outputAll = []

  ##### extract checkboxes
  rects = page.rects
  # Getting all the rect objects that are of square shape and in the range of 10-15 units per side and having no stroking color
  checkboxesALL = [rect for rect in rects 
                    if 5 < rect["width"] < 15 
                    and 5 < rect["height"] < 15 
                    and int(rect["width"]) == int(rect["height"]) 
                    and rect['non_stroking_color'] == None]
  im = page.to_image(resolution=200) 
  im.draw_lines(checkboxesALL)
  im.draw_rects(checkboxesALL)
  fn = os.path.join(path, "imageCB.png")
  im.save(fn, format="PNG")

  checkboxesChecked = []
  keysCBChecked = []
  # If a checkbox has a diagonal line within it, consider it checked.
  for checkbox in checkboxesALL:
    # Crop the page to contain just the checkbox rectangle.

    print(checkbox["x0"])
    print(checkbox["top"])    
    print(checkbox["x1"])
    print(checkbox["bottom"])

    cropped = page.crop((checkbox["x0"], checkbox["top"], checkbox["x1"], checkbox["bottom"]))      
    # If the cropped page has 6 edges, then consider it checked. Unchecked ones would have just 4 lines.

    print(len(cropped.edges))
    input("Press!")

    if len(cropped.edges) == 6:
      checkboxesChecked.append(checkbox)      
      keysCBChecked.append((checkbox["bottom"], checkbox["x0"]))
```

And this is the output i get:
```
239.39999999999998
289.0796120000001
244.43999999999997
294.2396120000001
16
```

So the program find the first checkbox (beside the nach § 19 EnEV text).
And also the cropping parameters looks correct to me.
But the cropped edges show me overall 16 edges for this object.
Why does this object has 16 edges - and why not 4 edges?

[sample1.pdf](https://github.com/jsvine/pdfplumber/files/10281542/sample1.pdf)

**Comment by jsvine** [comment]
> But the cropped edges show me overall 16 edges for this object. Why does this object has 16 edges - and why not 4 edges?

It looks like the PDF itself has drawn four separate rectangles in that part of the page, each with four edges. Not so uncommon for PDFs, which often include graphical objects that seem superfluous to humans. In any case, I'd suggest examining `cropped.rects` and determining which of the four will be most useful/consistent for your ultimate goals.

**Comment by Rapid1898-code** [comment]
Thanks a lot for the reply.

I saw for example this cropped.rects:

`{'x0': 239.39999999999998, 'y0': 558.920388, 'x1': 244.43999999999997, 'y1': 558.920388, 'width': 5.039999999999992, 'height': 0, 'pts': [(177.6, 558.920388), (422.88, 558.920388), (422.88, 541.280388), (177.6, 541.280388)], 'linewidth': 2.359, 'stroke': True, 'fill': False, 'evenodd': False, 'stroking_color': None, 'non_stroking_color': None, 'object_type': 'rect_edge', 'page_number': 1, 'top': 289.0796120000001, 'bottom': 289.0796120000001, 'doctop': 289.0796120000001, 'orientation': 'h'}`

`{'x0': 239.39999999999998, 'y0': 558.920388, 'x1': 244.43999999999997, 'y1': 558.920388, 'width': 5.039999999999992, 'height': 0, 'pts': [(234.23999999999998, 558.920388), (422.88, 558.920388), (422.88, 541.280388), (234.23999999999998, 541.280388)], 'linewidth': 2.359, 'stroke': True, 'fill': False, 'evenodd': False, 'stroking_color': None, 'non_stroking_color': None, 'object_type': 'rect_edge', 'page_number': 1, 'top': 289.0796120000001, 'bottom': 289.0796120000001, 'doctop': 289.0796120000001, 'orientation': 'h'}`

As you can see the have the same x0,y0,x1,y1 coordinates and height + width.
But the pts are completely different - but i have no idea what this information means.

**Reply by jsvine** [reply]
Looking at the items above, it seems that you're examining `.rect_edges` rather than `.rects`, because `'object_type': 'rect_edge'`. 

`pts` is an attribute that comes from `pdfminer.six` and isn't fully processed by `pdfplumber` (for instance, the distance-from-bottom vertical positions have not been converted into distance-from-top positions). For your purposes, I think you can ignore the `pts` attribute; examining `.rects`, however, I think will be more useful than `.rect_edges`.

**Comment by Rapid1898-code** [comment]
Hello - thanks again a lot for your help -
At the end i was able to identify if a checkbox is checked or not with the following code:

```
      if len(cropped.lines) == 2:
        checkboxesChecked.append(checkbox)
        keysCBChecked.append((checkbox["bottom"], checkbox["x0"]))     
```
With that i think it works for me.
